pkg/clusteragent/autoscaling/workload/model/pod_autoscaler.go

```diff
@@ -0,0 +1,244 @@
+type PodAutoscalerInternal struct {
+	// Namespace is the namespace of the PodAutoscaler
+	Namespace string
+
+	// Name is the name of the PodAutoscaler
+	Name string
+
+	// Generation is the received generation of the PodAutoscaler
+	Generation int64
+
+	// Keeping track of .Spec (configuration of the Autoscaling)
+	Spec datadoghq.DatadogPodAutoscalerSpec
+
+	// TargetGVK is the GroupVersionKind of the target resource
+	// Parsed once from the .Spec.TargetRef
+	TargetGVK schema.GroupVersionKind
+
+	// SettingsVersion is the version of the settings from RC
+	// (only if owner == remote)
+	SettingsVersion *uint64
+
+	// SettingsTimestamp is the time when the settings were last updated
+	// (only if owner == remote)
+	SettingsTimestamp time.Time
+
+	// Target represents the desired state
+	ScalingValues ScalingValues
+
+	// ScalingValuesVersion is the version of the values from RC
+	ScalingValuesVersion *uint64
+
+	// ScalingValuesTimestamp is the time when the values were last updated
+	ScalingValuesTimestamp time.Time
+
+	// HorizontalLastAction is the last horizontal action successfully taken
+	HorizontalLastAction *datadoghq.DatadogPodAutoscalerHorizontalAction
+
+	// HorizontalLastActionError is the last error encountered on horizontal scaling
+	HorizontalLastActionError error
+
+	// VerticalLastAction is the last action taken by the Vertical Pod Autoscaler
+	VerticalLastAction *datadoghq.DatadogPodAutoscalerVerticalAction
+
+	// VerticalRolloutError is the last error encountered on vertical scaling
+	VerticalRolloutError error
+
+	// CurrentReplicas is the current number of PODs for the targetRef
+	CurrentReplicas *int32
+
+	// ScaledReplicas is the current number of PODs for the targetRef matching the resources recommendations
+	ScaledReplicas *int32
+
+	// GlobalError is the an error encountered by the controller not specific to a scaling action
+	GlobalError error
+
+	// Deleted flags the PodAutoscaler as deleted (removal to be handled by the controller)
+	Deleted bool
+}
+
+// NewPodAutoscalerInternal creates a new PodAutoscalerInternal from a Kubernetes CR
+func NewPodAutoscalerInternal(_ string, podAutoscaler *datadoghq.DatadogPodAutoscaler) PodAutoscalerInternal {
+	pai := PodAutoscalerInternal{
+		Namespace:  podAutoscaler.Namespace,
+		Name:       podAutoscaler.Name,
+		Generation: podAutoscaler.Generation,
+		Spec:       podAutoscaler.Spec,
+	}
+	pai.UpdateFromStatus(podAutoscaler)
+	return pai
+}
+
+// ID returns the functional identifier of the PodAutoscaler
+func (p *PodAutoscalerInternal) ID() string {
+	return p.Namespace + "/" + p.Name
+}
+
+// UpdateFromSpec updates the PodAutoscalerInternal from a new spec
+func (p *PodAutoscalerInternal) UpdateFromSpec(podAutoscalerSpec *datadoghq.DatadogPodAutoscalerSpec) {
+	podAutoscalerSpec.DeepCopyInto(&p.Spec)
+	// Reset the target GVK as it might have changed
+	p.TargetGVK = schema.GroupVersionKind{}
```

Comment:
Because it can come from either update in Kubernetes or update from RC and we cannot report errors in these flows: we need to accept the input and report back the error in the status when we reconcile.

Reply:
I think it's worth adding a code comment about that